File: HW/HW12/PythonCode/WellCost.py
from typing import List


class WellCost:

    def findSet(target,setList):
        for i in range(len(setList)):
            set = setList[i]
            for numsInSet in set:
                for num in numsInSet:
                    if target == num:
                            return i
        return -1
    
    def minCostToSupplyWater(self, n: int, wells: List[int], pipes: List[List[int]]) -> int:
        edges = []
        sets = []
        cost = 0

        for house in range(len(wells)):
            edges.append([0,house + 1, wells[house]])
        
        for house in pipes:
            edges.append(house)
        
        edges.sort(key=lambda x: x[2])

        for set in edges:
            a = set[0]
            b = set[1]

            firstSetIndex = WellCost.findSet(a,sets)
            secondSetIndex = WellCost.findSet(b, sets)

            if firstSetIndex != -1 and secondSetIndex != -1 and firstSetIndex != secondSetIndex:
                sets[firstSetIndex].extend(sets[secondSetIndex])
                sets.pop(secondSetIndex)
                cost += set[2]

            elif firstSetIndex == -1 and secondSetIndex == -1:
                newSet = [[set[0],set[1]]]
                sets.append(newSet)
                cost += set[2]
            
            elif firstSetIndex != -1 and secondSetIndex == -1:
                sets[firstSetIndex].append([set[0],set[1]])
                cost += set[2]
            
            elif firstSetIndex == -1 and secondSetIndex != -1:
                sets[secondSetIndex].append([set[0],set[1]])
                cost += set[2]
            
            if len(sets) == 1 and len(sets[0]) == n:
                break
        
        return cost
    
test = WellCost()
houses = 2
wells = [1,1]
pipes = [[1,2,1], [1,2,2]]

print("Test 1: " + str(test.minCostToSupplyWater(houses,wells,pipes)))

houses = 3
wells = [3,3,3]
pipes = [[1,3,2],[1,1,4],[2,3,4]]

print("Test 2: " + str(test.minCostToSupplyWater(houses,wells,pipes)))

File: HW/HW12/JavaCode/WellCost.java
package HW.HW12.JavaCode;
import java.util.*;

public class WellCost{
    
    public int findSet(int target,ArrayList<ArrayList<int[]>> setList){
        for(int i = 0; i < setList.size(); i++){
            ArrayList<int[]> set = setList.get(i);
            for(int[] numsInSet : set){
                for(int num : numsInSet){
                    if(target == num){
                        return i;
                    }
                }
            }
        }
        return -1;
    }

    public int minCostToSupplyWater(int n, int[] wells, int[][] pipes) {

        ArrayList<int[]> edges = new ArrayList<>();
        ArrayList<ArrayList<int[]>> sets = new ArrayList<>();
        int cost = 0;

        for(int house = 0; house < wells.length; house++){
            int[] x = {0, house + 1, wells[house]};
            edges.add(x);
        }
        
        for(int[] house : pipes){
            edges.add(house);
        }

        Collections.sort(edges, Comparator.comparingInt(a -> a[2]));

        for(int[] set : edges){
            int a = set[0];
            int b = set[1];

            int firstSetIndex = findSet(a, sets);
            int secondSetIndex = findSet(b, sets);

            if(firstSetIndex != -1 && secondSetIndex != -1 && firstSetIndex != secondSetIndex){ 
                sets.get(firstSetIndex).addAll(sets.get(secondSetIndex));
                sets.remove(secondSetIndex);
                cost += set[2];
            }

            else if(firstSetIndex == -1 && secondSetIndex == -1){ 
                ArrayList<int[]> newSet = new ArrayList<>();
                int[] tempSet = {set[0],set[1]};
                newSet.add(tempSet);
                sets.add(newSet);
                cost += set[2];
            }
            else if(firstSetIndex != -1 && secondSetIndex == -1){ 
                int[] tempSet = {set[0],set[1]};
                sets.get(firstSetIndex).add(tempSet);
                cost += set[2];
            }
            else if(firstSetIndex == -1 && secondSetIndex != -1){
                int[] tempSet = {set[0],set[1]};
                sets.get(secondSetIndex).add(tempSet);
                cost += set[2];
            }

            if(sets.size() == 1 && sets.get(0).size() == n){
                break;
            }
        }
                    
        return cost;
    }

    public static void main(String[] args) {
        int houses = 2;
        int[] wells = {1,1};
        int[][] pipes = {{1,2,1}, {1,2,2}};

        WellCost test = new WellCost();
        System.out.println("Test 1: " + test.minCostToSupplyWater(houses, wells, pipes)); 

        int houses2 = 3;
        int[] wells2 = {3,3,3};
        int[][] pipes2 =  {{1,3,2},{1,1,4},{2,3,4}};
        System.out.println("Test 2: "+ test.minCostToSupplyWater(houses2, wells2, pipes2)); 
    }
}